importer/balanced/builder.go

```diff
@@ -370,3 +373,292 @@ func fillNodeRec(db h.DagBuilderHelperInterface, node *h.FSNodeOverDag, depth in
+		// Add the `filledBase` as a child of the `newRoot`.
+		newRoot := db.NewFSNodeOverDag(fsType)
+		err = newRoot.AddChild(filledBase, fileSize, db)
+		if err != nil {
+			return nil, err
+		}
+
+		// Fill out the sub-DAG topped by `filledBase`.
+		filledBase, fileSize, err = fillNodeRec(db, newRoot, currDepth, fsType)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return filledBase, db.Add(filledBase)
+}
+
+// VerifyParams is used by VerifyBalancedDagStructure
+type VerifyParamsForBalanced struct {
+	Getter    ipld.NodeGetter
+	MaxLinks  int
+	TreeDepth int
+	Prefix    *cid.Prefix
+	RawLeaves bool
+	Metadata  bool
+}
+
+// VerifyBalancedDagStructure checks that the given dag matches
+// exactly the balanced dag datastructure layout
+func VerifyBalancedDagStructure(nd ipld.Node, p VerifyParamsForBalanced) error {
+	return verifyBalancedDagRec(nd, p)
+}
+
+// Recursive call for verifying the structure of a balanced dag
+func verifyBalancedDagRec(n ipld.Node, p VerifyParamsForBalanced) error {
+	codec := cid.DagProtobuf
+	depth := p.TreeDepth
+	if depth == 0 {
+		if len(n.Links()) > 0 {
+			return errors.New("expected direct block")
+		}
+		// zero depth dag is raw data block
+		switch nd := n.(type) {
+		case *dag.ProtoNode:
+			fsn, err := ft.FSNodeFromBytes(nd.Data())
+			if err != nil {
+				return err
+			}
+
+			if fsn.Type() != ft.TFile && fsn.Type() != ft.TRaw && fsn.Type() != ft.TTokenMeta {
+				return errors.New("expected data or raw block or metadata block")
+			}
+
+			if p.RawLeaves {
+				return errors.New("expected raw leaf, got a protobuf node")
+			}
+		case *dag.RawNode:
+			if !p.RawLeaves {
+				return errors.New("expected protobuf node as leaf")
+			}
+			codec = cid.Raw
+		default:
+			return errors.New("expected ProtoNode or RawNode")
+		}
+	}
+
+	// verify prefix
+	if p.Prefix != nil {
+		prefix := n.Cid().Prefix()
+		expect := *p.Prefix // make a copy
+		expect.Codec = uint64(codec)
+		if codec == cid.Raw && expect.Version == 0 {
+			expect.Version = 1
+		}
+		if expect.MhLength == -1 {
+			expect.MhLength = prefix.MhLength
+		}
+		if prefix != expect {
+			return fmt.Errorf("unexpected cid prefix: expected: %v; got %v", expect, prefix)
+		}
+	}
+
+	if depth == 0 {
+		return nil
+	}
+
+	nd, ok := n.(*dag.ProtoNode)
+	if !ok {
+		return errors.New("expected ProtoNode")
+	}
+
+	// Verify this is a branch node
+	fsn, err := ft.FSNodeFromBytes(nd.Data())
+	if err != nil {
+		return err
+	}
+
+	if p.Metadata {
+		if fsn.Type() != ft.TTokenMeta {
+			return fmt.Errorf("expected token meta as branch node, got: %s", fsn.Type())
+		}
+	} else {
+		if fsn.Type() != ft.TFile {
+			return fmt.Errorf("expected file as branch node, got: %s", fsn.Type())
+		}
+	}
+
+	if fsn.Type() == ft.TFile && len(fsn.Data()) > 0 {
+		return errors.New("branch node should not have FS data")
+	}
+
+	for i := 0; i < len(nd.Links()); i++ {
+		child, err := nd.Links()[i].GetNode(context.TODO(), p.Getter)
```

Comment:
Please pass from upper level, or somewhere that already has a context.

Reply:
dm.GetCtx() - maybe

Reply:
Done.